cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java

```diff
@@ -210,42 +208,23 @@ private void addColumns(Connection connection, StructuredTableSchema schema)
   }
 
   // Add new non-primary key indices to table
-  private void addIndices(Connection connection, StructuredTableSchema schema)
+  private void addIndices(Connection connection, StructuredTableSchema newSchema, StructuredTableSchema oldSchema)
     throws IOException, TableNotFoundException {
     try (Statement statement = connection.createStatement()) {
       // Create indexes
-      for (String indexStatement : getCreateIndexStatements(schema.getTableId(),
-                                                            getNonExistIndexes(connection, schema))) {
+      Set<String> newIndexes = Sets.difference(newSchema.getIndexes(), oldSchema.getIndexes());
+      for (String indexStatement : getCreateIndexStatements(newSchema.getTableId(), newIndexes)) {
         LOG.debug("Create index statement: {}", indexStatement);
         statement.execute(indexStatement);
       }
     } catch (SQLException e) {
       if (TABLE_UNDEFINED_SQL_STATE.equalsIgnoreCase(e.getSQLState())) {
-        throw new TableNotFoundException(schema.getTableId());
+        throw new TableNotFoundException(newSchema.getTableId());
       }
-      throw new IOException(String.format("Error adding indices to table: %s", schema.getTableId().getName()), e);
+      throw new IOException(String.format("Error adding indices to table: %s", newSchema.getTableId().getName()), e);
     }
   }
 
-  private Set<String> getNonExistIndexes(Connection connection,
-                                         StructuredTableSchema schema) throws SQLException {
-    DatabaseMetaData metaData = connection.getMetaData();
-    Set<String> existingIndexes = new HashSet<>();
-    try (ResultSet rs = metaData.getIndexInfo(null, null,
```

Comment:
Remove unnecessary DB queries to construct columns and indexes, since we have them loaded in cache